text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+    }
+
+    fn build(self) -> Foo {
+        Foo {
+            alpha: self.alpha.unwrap(),
+            beta: self.beta.unwrap(),
+            gamma: self.gamma.unwrap_or(0),
+        }
+    }
+}
+```
+
+```rust
+pub struct Foo {
+    pub alpha: &'static str,
+    pub beta: bool,
+    pub gamma: i32,
+}
+
+// A builder type that is able to construct a `Foo`, but that will fail at compile time if a field
+// is missing.
+#[derive(Default)]
+pub struct FooBuilder<const A: bool, const B: bool, const G: bool> {
+    alpha: Option<&'static str>,
+    beta: Option<bool>,
+    gamma: Option<i32>,
+}
+
+// We provide this `impl` on its own so that `FooBuilder::new()` will work without specifying the
+// const parameters.
+impl FooBuilder<false, false, false> {
+    fn new() -> FooBuilder<false, false, false> {
+        FooBuilder::default()
+    }
+}
+
+// The fields can only be set once. Calling `set_alpha` twice will result in a compilation error.
+impl<const B: bool, const G: bool> FooBuilder<false, B, G> {
+    fn set_alpha(mut self, alpha: &'static str) -> FooBuilder<true, B, G> {
+        self.alpha = Some(alpha);
+        unsafe { std::mem::transmute(self) }
+    }
+}
+impl<const A: bool, const G: bool> FooBuilder<A, false, G> {
+    fn set_beta(mut self, beta: bool) -> FooBuilder<A, true, G> {
+        self.beta = Some(beta);
+        unsafe { std::mem::transmute(self) }
+    }
+}
+impl<const A: bool, const B: bool> FooBuilder<A, B, false> {
+    fn set_gamma(mut self, gamma: i32) -> FooBuilder<A, B, true> {
+        self.gamma = Some(gamma);
+        unsafe { std::mem::transmute(self) }
+    }
+}
+// If any field is optional, 
+impl<const G: bool> FooBuilder<true, true, G> {
+    fn build(self) -> Foo { // can only be called if all fields have been set
+        Foo {
+            alpha: self.alpha.unwrap(),
+            beta: self.beta.unwrap(),
+            gamma: self.gamma.unwrap_or(0), // This is an optional field with a default.
+        }
+    }
+}
+
+fn main() {
+    let _ = FooBuilder::new()
+        .set_alpha("")
+        .set_beta(false) // If we comment this out, it will no longer compile.
+        .set_gamma(42) // If we comment this out, it will still compile.
+        .build();
+}
+```
+
+All of the above can be represented with the exact same results with struct
+field default values, but with much less boilerplate:
+
+```rust
+pub struct Foo {
+    pub alpha: &'static str,
+    pub beta: bool,
+    pub gamma: i32 = 0,
+}
+
+fn main() {
+    let _ = Foo {
+        alpha: "",
+        beta: false,
+        ..
+    };
+}
+```
+
+The builder pattern is quite common in the Rust ecosystem, but as shown above its need is greatly
+reduced with `struct` field defaults.
+
+### `PhantomData` ergonomics
+
+[nomicon]: https://doc.rust-lang.org/nightly/nomicon/phantom-data.html
+[PhantomData]: https://doc.rust-lang.org/stable/std/marker/struct.PhantomData.html
+
+Say that you want to define our own `Vec<T>` type. As the [nomicon] says,
```

Comment:
The nomicon [no longer says this](https://github.com/rust-lang/nomicon/pull/363).